minio/minio
# config: migrate to better-suited encryption scheme

## Description
This commit implements a configuration migration for
the backend config file and IAM users/policies (`/.minio.sys/config/*`)
to a better-suited encryption scheme.

The main problem of the previous scheme was that it was
not designed to protect more then a few (e.g. the config file itself)
backend objects.
In particular, for each encrypted object an unique key was derived from
the S3 access key and secret key of the admin via a PBKDF - i.e.
Argon2id.

Argon2id is a memory-hard function designed to be computational
expensive. However, this would not have been a problem if only
a few objects would have been encrypted. However, the scheme was
used to also encrypt IAM users. If a cluster has many IAM users
(e.g. 1000) (re)starting the cluster nodes would take a significant
time since each node would need to perform 1000 or more Argon2id
executions. This lead to startup times of several seconds or even
minutes.

***

The new scheme addresses the issue of node startup time by migrating
to a (slightly) different but better suited encryption scheme.
The main difference is that the new scheme uses Argon2id to generate a
master key from the S3 secret key. This master key is then used to
derive unique encryption keys for each config/IAM user object.
However, these encryption keys are derived via a KDF/PRF (i.e.
HMAC-SHA256) which is much cheaper then a memory-hard PBKDF - like
Argon2id.

So, instead of deriving a new encryption key from the S3 secret key
using Argon2id per object, we now derive a master key via Argon2id
and then derive encryption keys from this master key via HMAC-SHA-256.

This reduces the number of Argon2id operations for `n` encrypted objects
to `1` instead of `n`. So, you perform an Argon2id execution once
regardless whether you have 1 or 10000 IAM users.

***

To achieve this the new scheme has to remember the key derivation
parameters (salt, time cost, memory cost, ...) somewhere. This
commit uses the `backend-encrypted` marker object for this purpose.

A `backend-encrypted` object for the new format looks similar to:
```
{
  "status": "encrypted",
  "derivation_params": {
    "algorithm": "Argon2id",
    "salt":      "sPx2HKZfuPsJc6OFrpIOIw==",
    "time":      1,
    "memory":    65536,
    "threads":   4
  }
}
```

## Motivation and Context
Server config / Startup time

## How to test this PR?
1. Run any MinIO version:
    ```
    export MINIO_ACCESS_KEY=minio
    export MINIO_SECRET_KEY=minio123
    minio server /tmp/1
    ```
2. Use the MinIO version with the change and restart
    ```
     minio server /tmp/1
     ```

Note that there are also cases like - no config encryption (access/secret key are not set as env. var.) and server startup is interrupted. 
   
## Types of changes

## Checklist:
